File: 백준/Gold/2812. 크게 만들기/크게 만들기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;
public class Main {
     public static void main(String[] args) throws IOException {
         BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
         StringTokenizer st = new StringTokenizer(br.readLine());
         int N = Integer.parseInt(st.nextToken());
         int K = Integer.parseInt(st.nextToken());
         String nums = br.readLine();
         Stack<Integer> stack = new Stack<>();
         int idx = 0;
         while (idx < N) {
             while (!stack.isEmpty()) {
                 if (stack.peek() < nums.charAt(idx) - '0') {
                     stack.pop();
                     K--;
                     if (K == 0) {
                         break;
                     }
                 }
                 else {
                     break;
                 }
             }
             if (K == 0) {
                 break;
             }
             stack.push(nums.charAt(idx) - '0');
             idx++;
         }

         while (idx < N) {
             stack.push(nums.charAt(idx) - '0');
             idx++;
         }

         while (K > 0) {
             stack.pop();
             K--;
         }
         StringBuilder sb = new StringBuilder();
         for (int i = 0; i < stack.size(); i++) {
             sb.append(stack.get(i));
         }
         System.out.println(sb);
     }
}

File: 백준/Gold/2812. 크게 만들기/크게 만들기.py
N, K = map(int, input().split())
num = list(input())
stack = []
idx = 0
while idx < N:
    while stack:
        if stack[-1] < num[idx]:
            stack.pop()
            K -= 1
            if K == 0:
                break
        else:
            break
    if K == 0:
        break
    stack.append(num[idx])
    idx += 1

while idx < N:
    stack.append(num[idx])
    idx += 1

while K > 0:
    stack.pop()
    K -= 1

print(''.join(map(str, stack)))

'''
6 2
999899
'''